Add `repo list` command

cristiand391 opened the pull request:
Closes #642 

This PR adds support for listing repositories of a user or organization, some examples:

TTY Output
![Screenshot_2021-02-25_12-08-30](https://user-images.githubusercontent.com/6853656/109173208-815d7300-777b-11eb-95b2-10b65b4a05e8.png)


Non-TTY Output
![Screenshot_2021-02-25_12-09-01](https://user-images.githubusercontent.com/6853656/109173228-891d1780-777b-11eb-98d5-f95a29118f04.png)

UI Spec: https://docs.google.com/document/d/1QoktdMTS8FEXYLH_1WvzZe_OoGV_ZiTxf0c3QVF1soY/edit#

cristiand391 committed:
Add `repo list` command

cristiand391 commented on pkg/cmd/repo/list/http.go:
```diff
@@ -0,0 +1,176 @@
+	if hasArchivedFilter {
+		// If `--archived` was specified, use the `search` API rather than
+		// `repositoryOwner.repositories`
+		query = fragment + `
```
[`repositoryOwner.repositories`](https://docs.github.com/en/graphql/reference/interfaces#repositoryowner) doesn't allow to filter repositories based on whether they are archived or not, the only way I found to do that was through the Search API.

Caveat:
It returns an ordered list but the `updatedAt` field isn't consistent with the one provided by `repositoryOwner` and makes the list look like if it were unordered.

`gh`

![Screenshot_2021-02-11_22-14-11](https://user-images.githubusercontent.com/6853656/107719384-0783af00-6cd0-11eb-90ed-23269d4def88.png)

`GitHub web`

![Screenshot_2021-02-11_22-20-38](https://user-images.githubusercontent.com/6853656/107719640-a01a2f00-6cd0-11eb-8fc5-58145efebd03.png)

cristiand391 commented on pkg/cmd/repo/list/list.go:
```diff
@@ -0,0 +1,224 @@
+		if notArchived && repo.IsArchived {
+			matchCount--
+			listResult.TotalCount--
+			continue
```
When doing `gh repo list --fork` or `gh repo list --source`, don't list archived repos.
This mimics the way GitHub does filter repos in the web UI.

mislav commented:
@cristiand391 Thanks for picking this up! This looks great already.

I'm going to suggest some output tweaks that are different from the design spec you were working off of, and ask input from my team to see if any of this resonates with them:

- I'm not sure if we need the `--archived` filter. Can anyone think of a use-case for it? I don't think we need to necessarily duplicate every filtering option available in the web UI— at least not for now.
- “When doing gh repo list `--fork` or gh repo list `--source`, don't list archived repos.” - Even if the web UI does this (TIL!), should we also do it? Since this command will have programmatic use, maybe it's best to avoid encoding some magic behavior. If we anticipate the need for avoiding archived repos, we could expose a flag such as `--no-archived`?
- Could repo tags such as “Private, archived" be rendered into the 3rd column (instead of the 2nd), _not_ be placed in parentheses (we only do this for repo labels), and be separated by commas? This way, the first 2 columns will be repo name & title, which are in my view the most important information.

ampinsk commented:
> I'm not sure if we need the --archived filter. Can anyone think of a use-case for it? I don't think we need to necessarily duplicate every filtering option available in the web UI— at least not for now.

I'm honestly agnostic on this one! If there's no strong push for it, I'm fine to punt on it.

> “When doing gh repo list --fork or gh repo list --source, don't list archived repos.” - Even if the web UI does this (TIL!), should we also do it? Since this command will have programmatic use, maybe it's best to avoid encoding some magic behavior. If we anticipate the need for avoiding archived repos, we could expose a flag such as --no-archived?

I don't feel super strongly about this either. But, even for programatic use cases, I can imagine if you archive a repository, it might be a lot of clutter to see them showing up in lists by default? And as a get around, wouldn't passing both `--fork` and `--archived` show archived forks for example?

> Could repo tags such as “Private, archived" be rendered into the 3rd column (instead of the 2nd), not be placed in parentheses (we only do this for repo labels), and be separated by commas? This way, the first 2 columns will be repo name & title, which are in my view the most important information.

I like this change! 👍

billygriffin commented:
> I'm not sure if we need the --archived filter. Can anyone think of a use-case for it? I don't think we need to necessarily duplicate every filtering option available in the web UI— at least not for now.

+1. I definitely don't think this is needed unless we actually see some significant demand for it.

> When doing gh repo list --fork or gh repo list --source, don't list archived repos.” - Even if the web UI does this (TIL!), should we also do it? Since this command will have programmatic use, maybe it's best to avoid encoding some magic behavior. If we anticipate the need for avoiding archived repos, we could expose a flag such as --no-archived?

I have no strong opinions about this either. Honestly I don't think it matters that much, and therefore would opt to mirror what dotcom does, especially since that helps reduce clutter as @ampinsk pointed out.

> Could repo tags such as “Private, archived" be rendered into the 3rd column (instead of the 2nd), not be placed in parentheses (we only do this for repo labels), and be separated by commas? This way, the first 2 columns will be repo name & title, which are in my view the most important information.

This is 💯

cristiand391 committed:
Remove archived filter from repo list

cristiand391 committed:
repo list: render repo tags into the 3rd column instead of the 2nd

cristiand391 committed:
repo list: add tests

cristiand391 commented:
@mislav I've added some tests and pushed some changes to address suggestions.
Regarding to listing archived repos when filtering, should I remove it or add some new flag or something else?

mislav committed:
Switch `repo list` to query via `graphql` package

Also order results by PUSHED_AT instead of UPDATED_AT to match the web
interface.

mislav committed:
Avoid having to first query for username in `repo list`

Dynamically construct the GraphQL query by using the `viewer` connection
if the owner isn't set and the `repositoryOwner(login:"...")` connection
if the owner was set.

mislav committed:
Isolate flag processing tests in `repo list`

mislav reviewed (approved):
Thank you for all your work!

I've pushed changes to:
1. Avoid having to separately request the current username when owner isn't specified via argument;
2. ~Remove all logic that filters out archived repos, since this level of filtering isn't supported in the API layer~;
3. Sort the returned results by `PUSHED_AT` instead of `UPDATED_AT` to match the web interface;
4. Enable the pager for tty output.

mislav committed:
Enable pager for `repo list` output

data-man commented:
Feature request: filtering by language.
Currently I'm using this tiny script:
```shell
set -e

repos() {
  local owner="${1?}"
  shift 1
  gh api graphql --paginate -f owner="$owner" "$@" -f query='
    query($owner: String!, $per_page: Int = 100, $endCursor: String) {
      repositoryOwner(login: $owner) {
        repositories(first: $per_page, after: $endCursor, ownerAffiliations: OWNER) {
          nodes {
            nameWithOwner
            description
            primaryLanguage { name }
            isFork
            pushedAt
          }
          pageInfo {
            hasNextPage
            endCursor
          }
        }
      }
    }
  ' | jq -r '.data.repositoryOwner.repositories.nodes[] | [.nameWithOwner,.pushedAt,.description,.primaryLanguage.name,.isFork] | @tsv' | sort
}

repos "$@"
```

mislav committed:
Enable filtering `repo list` by coding language

mislav commented:
@data-man That's an excellent suggestion. Added!

Note that for filtering by language, we have to turn to the `search` API connection (as opposed to the `repositories` connection). In that case, the results are sorted in `updated-desc` order rather than in `PUSHED_AT` order.

mislav committed:
Add ability to filter by archived in `repo list`

Like `--language`, archived filters also use the Search API.

data-man commented:
Proposals:
```
-a      --archived          Show only archived repositories
-f      --fork              Show only forks
-n      --no-archived       Omit archived repositories
-s      --source            Show only non-forks
```

mislav commented:
@data-man Thanks, but I do not think that every long flag needs to have a shorthand. In particular, `-f` and `-s` don't read really well when on their own, as opposed to `-l ruby` which is likely self-evident. Therefore, I'd rather that people spell out `--fork` and `--source`.

data-man commented:
@mislav 

It would be nice to show a primary language as well.

Gpapi13 reviewed (commented):
N and P

bruna17 reviewed (commented):
<details>
<summary>Details</summary>



</details>

cuddytrippin reviewed (commented):
> pull requests  #3061 `
